src/core/internal/arrayop.d

```diff
@@ -224,7 +283,7 @@ string scalarExp(Args...)()
         }
         else static if (isBinaryOp(arg))
         {
-            stack[$ - 2] = "(cast(T)(" ~ stack[$ - 2] ~ " " ~ arg ~ " " ~ stack[$ - 1] ~ "))";
```

Comment:
The existing arrayop behavior only casts the whole RHS before assignment, but not intermediate expressions. This is fixed here.